src/lib/WalletInfoCollector.ts

```diff
@@ -0,0 +1,217 @@
+
+        if (walletType === WalletType.BIP39 && WalletInfoCollector._keyguardClient) {
+            WalletInfoCollector._keyguardClient.releaseKey(walletId);
+        }
+
+        await initialAccountsPromise; // make sure initial accounts balances are updated
+        return walletInfo;
+    }
+
+    // TODO move to constants
+    private static readonly MAX_ALLOWED_GAP = 20;
+    private static readonly KEYGUARD_SLIP0010_BASE_PATH = `m/44'/242'/0'/`;
+    private static readonly DEFAULT_LABEL_LEGACY_WALLET = 'Legacy Wallet';
+    private static readonly DEFAULT_LABEL_KEYGUARD_WALLET = 'Keyguard Wallet';
+    private static readonly DEFAULT_LABEL_LEDGER_WALLET = 'Ledger Wallet';
+    private static readonly DEFAULT_LABEL_KEYGUARD_ACCOUNT = 'Standard Account';
+    private static readonly DEFAULT_LABEL_LEDGER_ACCOUNT = 'Ledger Account';
+    private static _keyguardClient?: KeyguardClient; // TODO avoid the need to create another KeyguardClient here
+    private static _networkInitializationPromise?: Promise<void>;
+    private static _wasmInitializationPromise?: Promise<void>;
+
+    private static _initializeDependencies(walletType: WalletType): void {
+        WalletInfoCollector._networkInitializationPromise =
+            WalletInfoCollector._networkInitializationPromise || NetworkClient.Instance.init();
+        WalletInfoCollector._networkInitializationPromise
+            .catch(() => delete WalletInfoCollector._networkInitializationPromise);
+        if (walletType === WalletType.BIP39) {
+            WalletInfoCollector._keyguardClient = WalletInfoCollector._keyguardClient || new KeyguardClient();
+        } else if (walletType === WalletType.LEDGER) {
+            WalletInfoCollector._wasmInitializationPromise =
+                WalletInfoCollector._wasmInitializationPromise || Nimiq.WasmHelper.doImportBrowser();
+            WalletInfoCollector._wasmInitializationPromise
+                .catch(() => delete WalletInfoCollector._wasmInitializationPromise);
+        }
+    }
+
+    private static async _getWalletInfoInstance(walletId: string, walletType: WalletType): Promise<WalletInfo> {
+        const existingWalletInfo = await WalletStore.Instance.get(walletId);
+        if (existingWalletInfo) return existingWalletInfo;
+        const label = walletType === WalletType.LEGACY
+            ? WalletInfoCollector.DEFAULT_LABEL_LEGACY_WALLET
+            : walletType === WalletType.BIP39
+                ? WalletInfoCollector.DEFAULT_LABEL_KEYGUARD_WALLET
+                : WalletInfoCollector.DEFAULT_LABEL_LEDGER_WALLET;
+        return new WalletInfo(
+            walletId,
+            label,
+            new Map<string, AccountInfo>(),
+            [],
+            walletType,
+            false,
+        );
+    }
+
+    private static async _deriveAccounts(startIndex: number, count: number, walletType: WalletType, walletId?: string)
+        : Promise<BasicAccountInfo[]> {
+        if (walletType === WalletType.LEGACY) {
```

Comment:
This is 4 cases for a variable check.
At that point i usually change it to a switch statement, even more so if every case either returns or throws.

Again, this is only personal preference. I am interested in your opinion mostly.

Reply:
Yes, I agree a switch here would look better.